src/assembler/index.ts

```diff
@@ -42,6 +42,11 @@ function createExecutable(objectFile: IELF): IELF {
       ...vectorTable.flatMap((x) => x.toBytes()),
       ...objectFile.content,
       ...END_OF_CODE.toBytes()
-    ]
+    ],
+    sourceMap: new Map<number, number>()
   }
+  objectFile.sourceMap.forEach((line, offset) => {
+    executable.sourceMap.set(FLASH_START.add(offset + 8).value, line)
+  })
+  return executable
```

Comment:
Maybe I do not see the whole picture but why do we again create an IELF object from the existing IELF object within the `createExecutable()` function? As far as I can see, the `vectorTable` and the `END_OF_CODE` are added. Is this just for better separation? Or couldn't the whole object straight away be build in the encoder module?

Reply:
It's just for better separation. And also I think it will make it easier when we are adding more features in the future.

ELF stands for executable and linkable format. So, basically it's a special format which can be used for linking and for executing. The `assembler` (in our case the `encoder.ts` file) has the job to take a code file and convert it into an *linkable* ELF file. So, it does have sections and symbol table and everything you would need for the linking process. On the other hand, the `linker` (in our case the `assembler/index.ts` file) has the job to convert it into an executable ELF file by adding a program header (segment information) which tells the loader how this should be loaded and executed.

Reply:
I will close this conversation. If you think you spotted an issue, do not hesitate to re-open it.